### ISSUE TITLE
[RFC] Introduce ATOM as external model package of SGLang

### ISSUE THREAD

**Issue by zhuyuhua-v** [ROCm/ATOM#359]
## Motivation

ATOM currently integrates with SGLang by maintaining a forked version of SGLang that adds `--model-impl atom` support. This approach requires invasive modifications to SGLang's codebase (6+ places), and imposes a continuous maintenance burden to keep the fork synchronized with upstream SGLang updates.

SGLang [PR #13429](https://github.com/sgl-project/sglang/pull/13429) introduced the `SGLANG_EXTERNAL_MODEL_PACKAGE` mechanism, which allows external packages to register custom model implementations **without modifying upstream SGLang**. This provides a clean, officially supported extension point that perfectly fits ATOM's use case.

By adopting this approach, we eliminate the need for a SGLang fork entirely. ATOM users can run optimized models on AMD GPUs using **unmodified upstream SGLang** with a single environment variable:

```bash
export SGLANG_EXTERNAL_MODEL_PACKAGE=atom.plugin.sglang.models
python3 -m sglang.launch_server --model-path <model_path> ...
```

### Comparison with the fork-based approach

| Dimension | `--model-impl atom` (fork) | `SGLANG_EXTERNAL_MODEL_PACKAGE` (now) |
|-----------|---------------------------|---------------------------------------|
| SGLang source | Modified fork, requires continuous sync | **Upstream SGLang, zero modifications** |
| Model registration | Built-in `ATOMForCausalLM` + `ModelImpl.ATOM` | External package with `EntryClass` |
| Supported models | Single wrapper for all architectures | Single base class, dynamic per-arch class generation |
| Maintenance cost | Only maintain ATOM-side code | Only maintain ATOM-side code |
| Adding new models | Append atom plugin registry | Append architecture name to `_MODEL_NAMES` list |

### SGLang's External Model Package Mechanism

SGLang's `SGLANG_EXTERNAL_MODEL_PACKAGE` ([PR #13429](https://github.com/sgl-project/sglang/pull/13429)) works through an automatic module discovery mechanism in `sglang/srt/models/registry.py`:

```
ModelRegistry = _ModelRegistry()
ModelRegistry.register("sglang.srt.models")                     # 1. Register built-in models

if external_pkg := envs.SGLANG_EXTERNAL_MODEL_PACKAGE.get():
    ModelRegistry.register(external_pkg, overwrite=True)         # 2. Register external models (overwrite)
```

When `SGLANG_EXTERNAL_MODEL_PACKAGE=atom.plugin.sglang.models` is set, the registry:

1. Imports the package (executes `__init__.py`)
2. Uses `pkgutil.iter_modules` to discover all `.py` submodules
3. Imports each submodule and collects classes from `EntryClass` attributes
4. Registers them by `cls.__name__`, with `overwrite=True` to replace built-in implementations

### ATOM Package Structure

```
atom/plugin/sglang/models/
└── base_model_wrapper.py   # Single base class + dynamic class generation for models
```

All currently supported model architectures are handled by a single file. `base_model_wrapper.py` defines:

- **`_AtomCausalLMBaseForSglang`** — a base `nn.Module` that conforms to SGLang's model interface and delegates to ATOM's plugin system
- **`_MODEL_NAMES`** — the currently enabled architecture names (`"DeepseekV3ForCausalLM"`, `"Qwen3MoeForCausalLM"`)
- **Dynamic class creation** — uses `type(name, (base,), {})` to generate concrete classes whose `__name__` matches the HF `architectures` field, then collects them into `EntryClass`

This design eliminates per-model boilerplate entirely: all wrapper classes share the same `__init__`, `forward`, and `load_weights` implementation from the base class.

### Architecture Diagram
<img width="1316" height="1078" alt="Image" src="https://github.com/user-attachments/assets/86937aed-a1a5-4e4e-8885-95d0489fddb8" />

### Execution Flow
<img width="1299" height="1122" alt="Image" src="https://github.com/user-attachments/assets/966cbd48-d8e7-44c5-bad6-cacdb854b41e" />

### Model Wrapper Implementation

All wrappers are generated from a single base class in `base_model_wrapper.py`:

```python
_MODEL_NAMES = [
    "DeepseekV3ForCausalLM",
    "Qwen3MoeForCausalLM",
]

class _AtomCausalLMBaseForSglang(nn.Module):
    def __init__(self, config, quant_config=None, prefix=""):
        super().__init__()
        self.pp_group = get_pp_group()
        self.config = config
        self.vocab_size = config.vocab_size
        self.unpadded_vocab_size = config.vocab_size

        import atom
        self.model = atom.prepare_model(config=config, engine="sglang")
        if self.model is None:
            model_arch = getattr(config, "architectures", ["unknown"])[0]
            raise ValueError(f"ATOM failed to create model for architecture {model_arch}")
        self.logits_processor = LogitsProcessor(config)

    @torch.no_grad()
    def forward(self, input_ids, positions, forward_batch,
                input_embeds=None, get_embedding=False,
                pp_proxy_tensors=None, **model_kwargs) -> Union[LogitsProcessorOutput, PPProxyTensors]:
        hidden_states = self.model(
            input_ids=input_ids,
            positions=positions,
            intermediate_tensors=pp_proxy_tensors,
            inputs_embeds=input_embeds,
            forward_batch=forward_batch,
            get_embedding=get_embedding,
            pp_proxy_tensors=pp_proxy_tensors,
            **model_kwargs,
        )
        if self.pp_group.is_last_rank:
            return self.logits_processor(
                input_ids, hidden_states, self.model.lm_head, forward_batch)
        return hidden_states

    def load_weights(self, weights):
        from atom.model_loader.loader import load_model_in_plugin_mode
        return load_model_in_plugin_mode(
            model=self.model, config=self.model.atom_config, prefix="model.")

# Dynamically generate concrete classes whose __name__ matches HF architectures
EntryClass = []
for _name in _MODEL_NAMES:
    _cls = type(_name, (_AtomCausalLMBaseForSglang,), {})
    globals()[_name] = _cls
    EntryClass.append(_cls)
```

Key design decisions:

- **Dynamic class generation via `type()`**: Instead of hand-writing per-model wrapper classes or per-model submodules, `type(name, (base,), {})` creates classes whose `__name__` matches the HF `architectures` field. This is what SGLang's `ModelRegistry` uses as the lookup key.
- **Weight loading via `load_model_in_plugin_mode`**: Rather than delegating to `self.model.load_weights(weights)`, the wrapper calls ATOM's `load_model_in_plugin_mode` directly. This function handles weight name mapping, sharding, quantization, and the `"model."` prefix stripping that aligns ATOM's internal weight names with the HF checkpoint layout.
- **Model validation**: After `atom.prepare_model()`, the wrapper checks for `None` return and raises a descriptive `ValueError` if ATOM doesn't support the requested architecture.

## Attention

SGLang provides the `register_attention_backend` decorator for registering custom attention backends. ATOM leverages this mechanism to inject its optimized attention implementation:

```python
@register_attention_backend("aiter")
def create_atom_backend(runner):
    from atom.plugin.sglang.attention_backend.sgl_attn_backend import ATOMAttnBackendForSgl
    return ATOMAttnBackendForSgl(runner)
```

The `ATOMAttnBackendForSgl` backend provides:
- **Prefill**: `flash_attn_varlen_func` from aiter for variable-length attention
- **Decode**: `pa_persistent_fwd` / `pa_fwd_asm` for paged attention with optimized AMD GPU kernels
- **KV Cache**: Shuffle-based layout via `reshape_and_cache_shuffle_triton`
- **CUDA Graph**: Support for capture and replay in decode phase

This registration uses the name `"aiter"` to align with SGLang's existing attention backend selection, ensuring the ATOM backend is transparently selected when running on AMD GPUs.

## Supported Models

| Model | Architecture Name | Source |
|-------|------------------|--------|
| DeepSeek-R1 / V3 | `DeepseekV3ForCausalLM` | `base_model_wrapper.py` |
| Qwen3 MoE (235B-A22B) | `Qwen3MoeForCausalLM` | `base_model_wrapper.py` |

All models are served by the same base class. The architecture name in `_MODEL_NAMES` determines which SGLang built-in class gets overridden.

## How to Add a New Model

Adding a new model to ATOM's package requires **no changes to upstream SGLang** and, for the common case, only a small ATOM-side update:

1. Add the model implementation to ATOM core (under `atom/models/`)
2. Register it in `atom/plugin/register.py`'s `_ATOM_SUPPORTED_MODELS`
3. Append the architecture name to `_MODEL_NAMES` in `base_model_wrapper.py`

For example, to add LLaMA support:

```python
_MODEL_NAMES = [
    "DeepseekV3ForCausalLM",
    "Qwen3MoeForCausalLM",
    "LlamaForCausalLM",        # ← add this line
]
```

No new files, no per-model wrapper classes, no environment variable changes. The dynamic `type()` loop generates a `LlamaForCausalLM` class that inherits the full base class behavior. SGLang's `pkgutil.iter_modules` discovers `base_model_wrapper.py`, finds the updated `EntryClass` list, and registers all classes including the new one.

Architectures that need wrapper-level special handling may require one extra hook. For example, DeepSeek V3 is also listed in `_DEEPSEEK_ARCHS` so the wrapper can run `setup_deepseek_for_sglang()` before inference.

## Usage

```bash
# Use upstream SGLang + ATOM via external model package
export PYTHONPATH=/path/to/upstream-sglang/python:/path/to/ATOM
export SGLANG_EXTERNAL_MODEL_PACKAGE=atom.plugin.sglang.models

python3 -m sglang.launch_server \
    --model-path DeepSeek-R1-0528 \
    --tensor-parallel-size 8 \
    --kv-cache-dtype fp8_e4m3 \
    --mem-fraction-static 0.8 \
    --page-size 1024 \
    --cuda-graph-max-bs 16
```

Switching models only requires changing `--model-path`. SGLang automatically selects the correct ATOM wrapper based on the model's HF `config.json` `architectures` field.

## Limitations

- Models not yet supported by ATOM's package will fall back to SGLang's built-in implementations. This fallback works correctly since the package only overrides specific architecture names.

## PRs

- SGLang external model support: [sgl-project/sglang#13429](https://github.com/sgl-project/sglang/pull/13429)
- ATOM implementation: [ROCm/ATOM#355](https://github.com/ROCm/ATOM/pull/355)
